File: app/src/main/java/com/itachi1706/busarrivalsg/util/StaticVariables.java
package com.itachi1706.busarrivalsg.util;

import android.content.SharedPreferences;
import android.util.Log;

import com.getpebble.android.kit.util.PebbleDictionary;
import com.itachi1706.busarrivalsg.objects.BusServices;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.Locale;
import java.util.UUID;
import java.util.concurrent.TimeUnit;

import javax.annotation.Nullable;

/**
 * Created by Kenneth on 20/6/2015
 * for SingBuses in package com.itachi1706.busarrivalsg
 */
public class StaticVariables {

    public static int HTTP_QUERY_TIMEOUT = 15000; //15 seconds timeout
    public static final String BASE_SERVER_URL = "http://api.itachi1706.com/api/appupdatechecker.php?action=androidretrievedata&packagename=";

    public final static UUID PEBBLE_APP_UUID = UUID.fromString("11198668-4e27-4e94-b51c-a27a1ea5cd82");

    public static final int CUR = 0, NEXT = 1, SUB = 2;

    public static final String USE_SERVER_TIME = "useServerTime";

    //For Pebble Comm
    public static PebbleDictionary dict1 = null;
    public static PebbleDictionary dict2 = null;
    public static PebbleDictionary dict3 = null;
    public static PebbleDictionary dict4 = null;

    public static ArrayList<BusServices> favouritesList = new ArrayList<>();

    public static boolean checkIfYouGotJsonString(String jsonString){
        return !jsonString.startsWith("<!DOCTYPE html>");
    }

    public static boolean useServerTime(SharedPreferences sp) {
        return sp.getBoolean(USE_SERVER_TIME, false);
    }

    public static long parseLTAEstimateArrival(String arrivalString, boolean useServerTime, @Nullable String serverTime) {
        if (arrivalString.equalsIgnoreCase("")) return -9999;
        return parseEstimateArrival(arrivalString, useServerTime, serverTime);
    }

    public static byte parseWABStatusToPebble(boolean status){
        if (status) return Byte.valueOf("1");
        return Byte.valueOf("0");
    }

    public static boolean checkBusLocationValid(double lat, double lng) {
        return !(lng == -1000 || lat == -1000) && !(lng == -11 && lat == -11) && !(lat == 0 && lng == 0);
    }

    public static long parseEstimateArrival(String arrivalString, boolean useServerTime, @Nullable String serverTime){
        Calendar currentDate;
        if (!useServerTime || serverTime == null) {
            Log.d("DATE", "Current Time Millis: " + System.currentTimeMillis());
            currentDate = Calendar.getInstance();
            currentDate.add(Calendar.MONTH, 1);
        } else {
            currentDate = parseDate(serverTime);
        }


        Calendar arrivalDate = parseDate(arrivalString);

        Log.d("COMPARE","Current: " + currentDate.toString() );
        Log.d("COMPARE","Arrival: " + arrivalDate.toString() );
        long difference = arrivalDate.getTimeInMillis() - currentDate.getTimeInMillis();
        return TimeUnit.MILLISECONDS.toMinutes(difference);
    }

    private static Calendar parseDate(String dateString) {
        Log.d("SPLIT", "Date String to parse: " + dateString);
        String[] firstSplit = dateString.split("T");
        String date = firstSplit[0];
        String time = firstSplit[1];
        String[] timeSplit = time.split("\\+");
        String trueTime = timeSplit[0];

        String[] dateSplit = date.split("-");
        int year = Integer.parseInt(dateSplit[0]);
        int month = Integer.parseInt(dateSplit[1]);
        int dates = Integer.parseInt(dateSplit[2]);

        String[] trueTimeSplit = trueTime.split(":");
        int hr = Integer.parseInt(trueTimeSplit[0]);
        int min = Integer.parseInt(trueTimeSplit[1]);
        int sec = Integer.parseInt(trueTimeSplit[2]);

        return new GregorianCalendar(year, month, dates, hr, min, sec);
    }

    public static String convertDateToString(Date date) {
        DateFormat df = new SimpleDateFormat("EE dd MMM yyyy HH:mm:ss zz", Locale.US);
        return df.format(date);
    }

    // HANDLER MESSAGES
    public static final int BUS_SERVICE_JSON_RETRIEVED = 101;
}

File: app/src/main/java/com/itachi1706/busarrivalsg/util/StaticVariables.kt
package com.itachi1706.busarrivalsg.util

import android.Manifest
import android.annotation.SuppressLint
import android.content.SharedPreferences
import com.itachi1706.busarrivalsg.objects.BusServices
import com.itachi1706.helperlib.helpers.LogHelper
import kotlinx.serialization.SerializationException
import kotlinx.serialization.json.Json
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.GregorianCalendar
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * Created by Kenneth on 20/6/2015
 * for SingBuses in package com.itachi1706.busarrivalsg
 */
object StaticVariables {

    const val BASE_SERVER_URL = "https://api.itachi1706.com/v1"

    const val CUR = 0
    const val NEXT = 1
    const val SUB = 2

    private const val USE_SERVER_TIME = "useServerTime"

    var favouritesList = ArrayList<BusServices>()

    // HANDLER MESSAGES
    const val BUS_SERVICE_JSON_RETRIEVED = 101

    @SuppressLint("CheckResult")
    fun checkIfYouGotJsonString(jsonString: String?): Boolean {
        if (jsonString == null) return false
        return try {
            Json.parseToJsonElement(jsonString)
            true
        } catch (_: SerializationException) {
            false
        }
    }

    fun useServerTime(sp: SharedPreferences): Boolean {
        return sp.getBoolean(USE_SERVER_TIME, false)
    }

    fun parseLTAEstimateArrival(
        arrivalString: String,
        useServerTime: Boolean,
        serverTime: String?
    ): Long {
        return if (arrivalString.equals("", ignoreCase = true)) -9999 else parseEstimateArrival(
            arrivalString,
            useServerTime,
            serverTime
        )
    }

    fun checkBusLocationValid(lat: Double, lng: Double): Boolean {
        return !(lng == -1000.0 || lat == -1000.0) && !(lng == -11.0 && lat == -11.0) && !(lat == 0.0 && lng == 0.0)
    }

    private fun parseEstimateArrival(
        arrivalString: String,
        useServerTime: Boolean,
        serverTime: String?
    ): Long {
        val currentDate: Calendar
        if (!useServerTime || serverTime == null) {
            LogHelper.d("DATE", "Current Time Millis: " + System.currentTimeMillis())
            currentDate = Calendar.getInstance()
            currentDate.add(Calendar.MONTH, 1)
        } else {
            currentDate = parseDate(serverTime)
        }


        val arrivalDate = parseDate(arrivalString)

        LogHelper.d("COMPARE", "Current: $currentDate")
        LogHelper.d("COMPARE", "Arrival: $arrivalDate")
        val difference = arrivalDate.timeInMillis - currentDate.timeInMillis
        return TimeUnit.MILLISECONDS.toMinutes(difference)
    }

    private fun parseDate(dateString: String): Calendar {
        LogHelper.d("SPLIT", "Date String to parse: $dateString")
        val firstSplit =
            dateString.split("T".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val date = firstSplit[0]
        val time = firstSplit[1]
        val timeSplit = time.split("\\+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val trueTime = timeSplit[0]

        val dateSplit = date.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val year = Integer.parseInt(dateSplit[0])
        val month = Integer.parseInt(dateSplit[1])
        val dates = Integer.parseInt(dateSplit[2])

        val trueTimeSplit =
            trueTime.split(":".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val hr = Integer.parseInt(trueTimeSplit[0])
        val min = Integer.parseInt(trueTimeSplit[1])
        val sec = Integer.parseInt(trueTimeSplit[2])

        return GregorianCalendar(year, month, dates, hr, min, sec)
    }

    fun convertDateToString(date: Date): String {
        val df = SimpleDateFormat("EE dd MMM yyyy HH:mm:ss zz", Locale.US)
        return df.format(date)
    }

    fun checkIfCoarseLocationGranted(result: Map<String, Boolean>): Boolean {
        return result[Manifest.permission.ACCESS_COARSE_LOCATION] ?: false
    }
}
